### PR TITLE
Add support for NO_PIN to all matrix types

### PR DESCRIPTION
**Pull request by drashna**
## Description

Adds support for `NO_PIN` in the regular matrices

## Types of Changes

- [x] Core
- [x] Enhancement/optimization

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/matrix.c [hunk 1]
```diff
@@ -107,8 +111,11 @@ static bool read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row)
 
     // For each col...
     for (uint8_t col_index = 0; col_index < MATRIX_COLS; col_index++) {
-        // Select the col pin to read (active low)
-        uint8_t pin_state = readPin(col_pins[col_index]);
+        uint8_t pin_state = 0;
```
Comment (fauxpark) [inline quantum/matrix.c:114]: If it's active low, should this not be 1?

**Comment by mtei** [comment]
I would say the better way is to define this function and use it.
```c
static inline uint8_t readMatrixPin(pin_t pin) {
   if (pin != NO_PIN) {
       retrun readPin(pin);
   } else {
       retrun 1;
   }
}
```

File: quantum/matrix.c (hunk 1 above)
Comment (drashna) [reply quantum/matrix.c:114]: It might be, yeah. :(

**Comment by mtei** [comment]
I have one more suggestion.

```diff
-static void select_row(uint8_t row) { setPinOutput_writeLow(row_pins[row]); }
+static bool select_row(uint8_t row) {
+    pin_t pin = row_pins[row];
+    if (pin != NO_PIN) {
+        setPinOutput_writeLow(pin);
+        return true;
+    }
+    return false;
+}
 
-static void unselect_row(uint8_t row) { setPinInputHigh_atomic(row_pins[row]); }
+static void unselect_row(uint8_t row) {
+    pin_t pin = row_pins[row];
+    if (pin != NO_PIN) {
+        setPinInputHigh_atomic(pin);
+    }
+}
 
 static void unselect_rows(void) {
     for (uint8_t x = 0; x < MATRIX_ROWS; x++) {
-        if (row_pins[x] != NO_PIN) {
-            setPinInputHigh_atomic(row_pins[x]);
-        }
+        unselect_row(x);
     }
 }
```

```diff
@@ -113,8 +123,9 @@ static bool read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row)
     // Start with a clear matrix row
     matrix_row_t current_row_value = 0;
 
-    // Select row
-    select_row(current_row);
+    if (!select_row(current_row)) { // Select row
+        return false;  // skip NO_PIN row
+    }
     matrix_output_select_delay();
 ```

File: quantum/matrix.c [hunk 2]
```diff
@@ -80,35 +88,50 @@ static bool read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row)
...
+    pin_t pin = row_pins[row];
+    if (pin != NO_PIN) {
+        setPinInputHigh_atomic(pin);
+    }
+}
 
 static void unselect_rows(void) {
     for (uint8_t x = 0; x < MATRIX_ROWS; x++) {
-        setPinInputHigh_atomic(row_pins[x]);
+        unselect_row(row_pins[x]);
```
Comment (tzarc) [inline quantum/matrix.c:109]:
```suggestion
        unselect_row(x);
```

File: quantum/matrix.c [hunk 3]
```diff
@@ -128,28 +151,44 @@ static bool read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row)
...
+    pin_t pin = col_pins[col];
+    if (pin != NO_PIN) {
+        setPinInputHigh_atomic(pin);
+    }
+}
 
 static void unselect_cols(void) {
     for (uint8_t x = 0; x < MATRIX_COLS; x++) {
-        setPinInputHigh_atomic(col_pins[x]);
+        unselect_col(col_pins[x]);
```
Comment (tzarc) [inline quantum/matrix.c:172]:
```suggestion
        unselect_col(x);
```

File: quantum/split_common/matrix.c [hunk 4]
```diff
@@ -94,35 +102,49 @@ static bool read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row)
...
+    pin_t pin = row_pins[row];
+    if (pin != NO_PIN) {
+        setPinInputHigh_atomic(pin);
+    }
+}
 
 static void unselect_rows(void) {
     for (uint8_t x = 0; x < ROWS_PER_HAND; x++) {
-        setPinInputHigh_atomic(row_pins[x]);
+        unselect_row(row_pins[x]);
```
Comment (tzarc) [inline quantum/split_common/matrix.c:123]:
```suggestion
        unselect_row(x);
```

File: quantum/split_common/matrix.c [hunk 5]
```diff
@@ -142,28 +164,44 @@ static bool read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row)
...
+    pin_t pin = col_pins[col];
+    if (pin != NO_PIN) {
+        setPinInputHigh_atomic(pin);
+    }
+}
 
 static void unselect_cols(void) {
     for (uint8_t x = 0; x < MATRIX_COLS; x++) {
-        setPinInputHigh_atomic(col_pins[x]);
+        unselect_col(col_pins[x]);
```
Comment (tzarc) [inline quantum/split_common/matrix.c:185]:
```suggestion
        unselect_col(x);
```

**Review by tzarc — CHANGES_REQUESTED** [review]

**Review by tzarc — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by elfmimi** [comment]
I need to leave a note here. I'm guessing that the reason support for NO_PIN didn't exist until today is because QMK takes binary side into account very seriously. I've heard QMK is turning its wheel to the direction of data driven design. If so, check for NO_PIN could have been made to activated only for those do include NO_PINs in the definition. This was my thought on this PR.

**Pull request merged**